### PR TITLE
Include series in harvest API (and improve lots of other stuff)

### PR DESCRIPTION
**Pull request by LukasKalbertodt**
Also see commit messages. Main open questions:

- Is the addition to the `SeriesSerivce` (and related stuff) ok like this? Will this be accepted in the community?
- ~~Do I really need to implement that method for the remote impl? Like... can we just say Tobira is only on the admin? Or sth like that?~~ We decided for now that we will just throw an exception and care about this later. 
- Are you happy with the new code and logic in the Tobira module?

The only things left, as far as I can tell, before calling this initial API finished:
- [ ] Adding good REST docs to it
- [ ] Adding more fields to the reply. Currently, hardly any useful information is given for event and series. That has to be more. But adding that should be very straight forward.
- [x] Renaming and correctly describing the `limit` parameter to clarify it's just a guideline and the API can return however many.
- [ ] DB migrations
- [ ] Index for `deletion_date`?

### CODE REVIEW

File: modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java [hunk 1]
```diff
@@ -373,7 +392,7 @@ public DublinCoreCatalog getSeries(String seriesId) throws NotFoundException, Se
     } catch (NotFoundException e) {
       throw e;
     } catch (Exception e) {
-      logger.error("Could not update series: {}", e.getMessage());
+      logger.error("Could not retrieve series with ID '{}': {}", seriesId, e.getMessage());
```
Comment (lkiesow) [inline modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java:395]:
```suggestion
      logger.error("Could not retrieve series with ID '{}", seriesId, e);
```

File: modules/tobira/src/main/java/org/opencastproject/tobira/impl/TobiraApi.java [hunk 2]
```diff
@@ -103,79 +138,154 @@ public void setSearchService(SearchService service) {
...
+        // that have a modified date after the one of the last raw series. But that means that we
+        // will return all series and events until the last raw series.
+        includesItemsUntilRaw = rawSeries.get(rawSeries.size() - 1).getModifiedDate();
+      } else if (hasMoreEvents && !hasMoreSeriesInRange) {
+        // There are more events, but no additional series in the range from `since` to the modified
+        // date of the last raw event. So we know there are no other events or series before the
+        // last raw events.
+        includesItemsUntilRaw = rawEvents[rawEvents.length - 1].getModified();
+      } else {
+        // There are more events and more series in the given range. In theory, this would be
```
Comment (lkiesow) [inline modules/tobira/src/main/java/org/opencastproject/tobira/impl/TobiraApi.java:258]: Not only related to this part of the code, but we spend a lot of effort on doing things to figure out if there are more events, … do we need that at all? Long-term, the harvesting should only be the fallback option, since we also want synchronous updates. Even if we don't know that there are more events, we will harvest again eventually. The limit is mostly for an initial harvest, so we do not load ten thousand events at once. But we know about the initial harvest and can have admins just increase the harvesting speed during that period if necessary.

So, this kind of feels a bit overengineered and kind of feels like code where someone may make the wrong turn eventually, causing an error.

File: modules/tobira/src/main/java/org/opencastproject/tobira/impl/TobiraApi.java [hunk 3]
```diff
@@ -103,79 +138,154 @@ public void setSearchService(SearchService service) {
...
+        // smaller than that of the last raw event.
+        includesItemsUntilRaw = rawSeries.get(rawSeries.size() - 1).getModifiedDate();
+      }
+
+      // The `includesItemsUntil` we return has to be at least `TIME_BUFFER_SIZE` in the past. See
+      // the constant's documentation for more information on that.
+      final long includesItemsUntil = Math.min(
+          includesItemsUntilRaw.getTime(),
+          new Date().getTime() - TIME_BUFFER_SIZE
+      );
```
Comment (lkiesow) [inline modules/tobira/src/main/java/org/opencastproject/tobira/impl/TobiraApi.java:271]: It's all documentation, but instead of modifying the value up until which this request includes data, it might be less confusing a) to modify the request parameter or b) expect the client to do that. That also makes this code easier since it would just become:
```suggestion
      final long includesItemsUntil = includesItemsUntilRaw.getTime();
```

File: modules/tobira/src/main/java/org/opencastproject/tobira/impl/TobiraApi.java [hunk 4]
```diff
@@ -103,79 +138,154 @@ public void setSearchService(SearchService service) {
...
+
+      // Assembly full response.
+      final List<Jsons.Val> outItems = items.stream()
+          .map(item -> item.getJson())
+          .collect(Collectors.toCollection(ArrayList::new));
       final Jsons.Obj json = Jsons.obj(
           Jsons.p("includesItemsUntil", includesItemsUntil),
           Jsons.p("hasMore", hasMore),
-          Jsons.p("items", Jsons.arr(items))
+          Jsons.p("items", Jsons.arr(outItems))
```
Comment (lkiesow) [inline modules/tobira/src/main/java/org/opencastproject/tobira/impl/TobiraApi.java:281]: _(Applied to all JSON related code)_

It's certainly personal preference, but I find it much easier to work with GSON instead of json-simple since you don't need to manually build all your JSON objects, which usually also makes the code much easier to read:

```
Series series = new Series();
…
String jsonString = gson.toJson(series);
```

or

```
List<Series> seriesList = new ArrayList<>();
…
String jsonString = gson.toJson(seriesList);
```

Of course, you can still build it manually, if you need to:
```
Map<String, Object> customData = new Map<>();
customData.put("key", "value");
String jsonString = gson.toJson(customData);
```

File: modules/tobira/src/main/resources/OSGI-INF/tobira-api.xml [hunk 5]
```diff
@@ -13,4 +13,7 @@
     <reference name="searchService"
                interface="org.opencastproject.search.api.SearchService"
                bind="setSearchService"/>
+    <reference name="seriesService"
```
Comment (lkiesow) [inline modules/tobira/src/main/resources/OSGI-INF/tobira-api.xml:16]: Can we replace the XML with OSGi declarative service annotations?

**Review by lkiesow — COMMENTED** [review]
Generally looks good. I did not notice any major problems when reading the code. I found a few smaller issues and added some suggestions for improvements. We don't have to include all of them.

One additional suggestion I think would make sense is to actually delete a series main data when deleting a series: Remove all elements and properties and set things like name or the DC catalog to null.

File: modules/tobira/src/main/resources/OSGI-INF/tobira-api.xml (hunk 5 above)
Comment (JulianKniephoff) [reply modules/tobira/src/main/resources/OSGI-INF/tobira-api.xml:16]: I actually do have some more understanding of the situation, and even a workaround. Talk to me if we want to pursue this. :wink:

Comment (lkiesow) [reply modules/tobira/src/main/resources/OSGI-INF/tobira-api.xml:16]: I don't want to know. I should bother to acquire that knowledge anyway …but  not today.

File: modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java [hunk 6]
```diff
@@ -383,6 +402,63 @@ public DublinCoreCatalog getSeries(String seriesId) throws NotFoundException, Se
...
+    }
+
+    // Load series from DB.
+    EntityManager em = emf.createEntityManager();
+    try {
+      TypedQuery<SeriesEntity> q;
+      if (to.isPresent()) {
+        if (from.after(to.get())) {
+          throw new IllegalArgumentException("`from` is after `to`");
+        }
```
[the remark this answers is not in the document]
Comment (LukasKalbertodt) [reply modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java:428]: The Tobira code does not benefit from this function returning an empty list. Thus I opted for throwing an error for now as this might well be a bug in the caller code.

File: modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java [hunk 7]
```diff
@@ -383,6 +402,63 @@ public DublinCoreCatalog getSeries(String seriesId) throws NotFoundException, Se
...
+            .setMaxResults(limit);
+      }
+
+      final List<Series> out = new ArrayList<>();
+      for (SeriesEntity entity : q.getResultList()) {
+        final Series series = new Series();
+        series.setId(entity.getSeriesId());
+        series.setOrganization(entity.getOrganization());
+        series.setDublinCore(
+            dcService.load(IOUtils.toInputStream(entity.getDublinCoreXML(), "UTF-8"))
```
[the remark this answers is not in the document]
Comment (LukasKalbertodt) [reply modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java:446]: Are you sure about this? It's done the same way in `getSeries`: https://github.com/elan-ev/opencast-tobira/blob/46ec3f27b2ea67dd9e02f0f9f55156af9ac7a929/modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java#L391

File: modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java [hunk 8]
```diff
@@ -534,13 +610,14 @@ public boolean storeSeriesAccessControl(String seriesId, AccessControlList acces
...
       entity.setAccessControl(serializedAC);
+      entity.setModifiedDate(new Date());
       em.merge(entity);
       tx.commit();
       return updated;
     } catch (NotFoundException e) {
       throw e;
     } catch (Exception e) {
-      logger.error("Could not update series: {}", e.getMessage());
+      logger.error("Could not store ACL for series '{}': {}", seriesId, e.getMessage());
```
[the remark this answers is not in the document]
Comment (LukasKalbertodt) [reply modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java]: Do you want me to change this for all these cases in the file? Basically every existing method first logs an error, then throws the exception. I can if you want, but it's completely unrelated to the rest of this PR and it might be better to immediately create such a PR against the community version.

File: modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesEntity.java [hunk 9]
```diff
@@ -49,17 +53,39 @@
...
+    @NamedQuery(
+        name = "Series.getAllModifiedSince",
+        query = "select s from SeriesEntity as s "
+            + "where s.modifiedDate >= :since "
+            + "order by s.modifiedDate asc"
+    ),
+    @NamedQuery(
+        name = "Series.getAllModifiedInRange",
+        query = "select s from SeriesEntity as s "
+            + "where s.modifiedDate >= :from and s.modifiedDate < :to "
```
[the remark this answers is not in the document]
Comment (LukasKalbertodt) [reply modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesEntity.java:86]: `from` alone sounds weird to me. And the pair `since` and `to` also sound weird. Which is why I opted for these names. But I don't really care. If you want, I can change `since` to `from`, for example.

File: modules/series-service-api/src/main/java/org/opencastproject/series/api/Series.java [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+  private Date deletionDate;
+
+
+  public String getId() {
+    return this.id;
+  }
+
+  public void setId(String id) {
+    this.id = id;
+  }
```
[the remark this answers is not in the document]
Comment (LukasKalbertodt) [reply modules/series-service-api/src/main/java/org/opencastproject/series/api/Series.java:60]: Is there precedence on these builder objects in OC? Is that something Java-people do often? If you think it would be better, I can change it. But it doesn't really benefit us here. 

I'm way more interested in just making all fields of the class public and getting rid of all methods entirely. But checkstyle won't allow it. Planning on starting a discussion about this in the matrix channel, once I find the energy to do so :P

File: modules/tobira/src/main/java/org/opencastproject/tobira/impl/TobiraApi.java (hunk 2 above)
Comment (LukasKalbertodt) [reply modules/tobira/src/main/java/org/opencastproject/tobira/impl/TobiraApi.java:258]: That's a very good point and that thought crossed my mind, too. If we keep the current logic, I certainly want to have tests for the various edge cases. And yes, the logic got a bit more involved than I anticipated. I was very close to just giving up on smart logic entirely, like you suggest.

However, right now I'm fairly happy with the code and the explaining comments. Not being smart about this at all does have a few disadvantages. It would mean transferring lots of duplicate series, sending stuff out of order very often (which Tobira needs to deal with anyway, BUT leads to temporary object and extra handling on Tobira side each time), and sending lots of items more than once. It's kind of nice if the harvesting can say "no new data" instead of always sending the last item again and again.

So yeah, I also have my troubles with the complexity this adds. Right now, I have a very small tendency to be smart here. Especially since we now already have the code and I don't expect to change it in the near future. But you'd certainly be able to convince me otherwise.

---

Final note about the synchronous API: There is still the idea in my head that Tobira merely offers one endpoint that, when requested, just queues a new harvest right away. Ala "hey, there is new data, please harvest now!". Sure, in the worst case, that would mean twice as many HTTP requests (though one would contain no data, i.e. is small). But this would be easy to add, would basically offer the same advantages as a "push data" API, and most importantly would avoid extra logic and complexities.

File: modules/tobira/src/main/java/org/opencastproject/tobira/impl/TobiraApi.java (hunk 3 above)
Comment (LukasKalbertodt) [reply modules/tobira/src/main/java/org/opencastproject/tobira/impl/TobiraApi.java:271]: I'm not sure I understand. The client just takes the `includesItemsUntil` value and uses it as the `since` value for the next harvest request. And I would say the `TIME_BUFFER_SIZE` should be decided by the Opencast side. So I'm not entirely sure what you're suggesting here.

File: modules/tobira/src/main/java/org/opencastproject/tobira/impl/TobiraApi.java (hunk 4 above)
Comment (LukasKalbertodt) [reply modules/tobira/src/main/java/org/opencastproject/tobira/impl/TobiraApi.java:281]: I am generally also a fan of these kinds of serialization libraries. On the Tobira side, we do it that way. But in this case I'm not sure if it would really buy us anything. As explained in private chat: to have complete control over the format of the API (which we should, IMO) we can't call `toJson` on the objects returned by the search and series API. Because those classes can be modified at any time, potentially breaking API compatibility. Thus, we need our own classes. But defining them, including fields, getters, setter AND the code to create them would add a lot of code to this module. So in this case, I think the json-simple works fine and results in less code. So I'd like to keep it for now.

File: modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (lkiesow) [reply modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java:446]: The DublinCoreService internally decides if it is JSON or XML, then (if it decides it is XML) calls DublinCoreXmlFormat. And DublinCoreXmlFormat  has a read method which accepts Strings directly, so you don't have to build an input stream.

File: modules/series-service-api/src/main/java/org/opencastproject/series/api/Series.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (lkiesow) [reply modules/series-service-api/src/main/java/org/opencastproject/series/api/Series.java:60]: - If I remember correctly, the Apache and Elastic modules or the AssetM;anager do that and it's essentially how Java's Streams work
- Most old code would probably expect a constructor with arguments to allow creating an object quickly without a lot of setter calls
- I'm a Python dev in heart → everything should be public. Don't know why anyone ever would want to make things private ;-P

File: modules/tobira/src/main/java/org/opencastproject/tobira/impl/TobiraApi.java (hunk 2 above)
Comment (lkiesow) [reply modules/tobira/src/main/java/org/opencastproject/tobira/impl/TobiraApi.java:258]: > It's kind of nice if the harvesting can say "no new data" instead of always sending the last item again and again.

Remind me again: Why would we need to send one item?

File: modules/series-service-api/src/main/java/org/opencastproject/series/api/Series.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (LukasKalbertodt) [reply modules/series-service-api/src/main/java/org/opencastproject/series/api/Series.java:60]: Okay interesting. 

> Not that important

So then lets leave it for now. I can still return later and change stuff.

**Comment by LukasKalbertodt** [comment]
We discusses that all remaining points can be addressed later. So I will merge as soon as CI is green.
